File: electron/agent/tools.ts
import fs from 'node:fs'
import path from 'node:path'
import { computeLineDiff, type DiffLine } from './diff'

export interface ToolContext {
  projectPath: string | null
  openFileContents: Map<string, string>
}

export interface ToolResult {
  ok: boolean
  output: string
  diff?: DiffLine[]
  filePath?: string
  filesChanged?: boolean
}

export const AGENT_TOOLS = [
  {
    type: 'function',
    function: {
      name: 'read_file',
      description: 'Read file contents. Use relative path from project root.',
      parameters: {
        type: 'object',
        properties: {
          path: { type: 'string', description: 'File path' },
          start_line: { type: 'number', description: 'Optional 1-based start line' },
          end_line: { type: 'number', description: 'Optional 1-based end line' },
        },
        required: ['path'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'edit_file',
      description: 'Replace exact text in a file. old_string must match exactly once.',
      parameters: {
        type: 'object',
        properties: {
          path: { type: 'string' },
          old_string: { type: 'string' },
          new_string: { type: 'string' },
        },
        required: ['path', 'old_string', 'new_string'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'write_file',
      description: 'Create or overwrite entire file.',
      parameters: {
        type: 'object',
        properties: {
          path: { type: 'string' },
          content: { type: 'string' },
        },
        required: ['path', 'content'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'list_directory',
      description: 'List files and folders in a directory.',
      parameters: {
        type: 'object',
        properties: { path: { type: 'string', description: 'Directory path, default "."' } },
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'search_codebase',
      description: 'Search for regex pattern in project files.',
      parameters: {
        type: 'object',
        properties: {
          pattern: { type: 'string' },
          glob: { type: 'string', description: 'Optional extension filter e.g. ts, py' },
        },
        required: ['pattern'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'summarize_file',
      description: 'Get file metadata and a short preview for large files.',
      parameters: {
        type: 'object',
        properties: { path: { type: 'string' } },
        required: ['path'],
      },
    },
  },
  {
    type: 'function',
    function: {
      name: 'web_search',
      description: 'Search the web for documentation, errors, or APIs.',
      parameters: {
        type: 'object',
        properties: { query: { type: 'string' } },
        required: ['query'],
      },
    },
  },
]

function resolvePath(ctx: ToolContext, filePath: string): string | null {
  if (!ctx.projectPath) return null
  if (filePath === '.' || filePath === './') return ctx.projectPath
  if (path.isAbsolute(filePath)) {
    const norm = path.normalize(filePath)
    if (!norm.startsWith(path.normalize(ctx.projectPath))) return null
    return norm
  }
  return path.normalize(path.join(ctx.projectPath, filePath))
}

function readTextFile(abs: string): { ok: true; content: string } | { ok: false; msg: string } {
  if (!fs.existsSync(abs)) return { ok: false, msg: 'File not found' }
  const buf = fs.readFileSync(abs)
  if (buf.includes(0)) return { ok: false, msg: 'Binary file — cannot read as text' }
  return { ok: true, content: buf.toString('utf8') }
}

function walkDir(dir: string, out: string[] = []): string[] {
  if (!fs.existsSync(dir)) return out
  for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
    if (e.name.startsWith('.') || e.name === 'node_modules') continue
    const p = path.join(dir, e.name)
    if (e.isDirectory()) walkDir(p, out)
    else out.push(p)
  }
  return out
}

async function webSearch(query: string): Promise<string> {
  try {
    const url = `https://api.duckduckgo.com/?q=${encodeURIComponent(query)}&format=json&no_redirect=1`
    const res = await fetch(url, { headers: { 'User-Agent': 'CelestiaAgent/1.0' } })
    if (!res.ok) return `Search failed: HTTP ${res.status}`
    const data = await res.json() as {
      Abstract?: string
      AbstractText?: string
      AbstractURL?: string
      RelatedTopics?: Array<{ Text?: string; FirstURL?: string } | { Topics?: { Text?: string }[] }>
    }
    const parts: string[] = []
    if (data.AbstractText) parts.push(data.AbstractText, data.AbstractURL ?? '')
    const topics = data.RelatedTopics ?? []
    for (const t of topics.slice(0, 5)) {
      if ('Text' in t && t.Text) parts.push(`• ${t.Text}`)
      if ('Topics' in t && t.Topics) {
        for (const sub of t.Topics.slice(0, 3)) {
          if (sub.Text) parts.push(`• ${sub.Text}`)
        }
      }
    }
    return parts.filter(Boolean).join('\n') || 'No results found.'
  } catch (err) {
    return `Search error: ${err instanceof Error ? err.message : String(err)}`
  }
}

export async function executeTool(
  name: string,
  args: Record<string, unknown>,
  ctx: ToolContext,
): Promise<ToolResult> {
  switch (name) {
    case 'read_file': {
      const abs = resolvePath(ctx, String(args.path ?? '').split(':')[0])
      if (!abs) return { ok: false, output: 'Invalid path or no project open' }
      const fromOpen = ctx.openFileContents.get(abs)
      const raw = fromOpen ?? (() => {
        const r = readTextFile(abs)
        return r.ok ? r.content : null
      })()
      if (raw === null) {
        const r = readTextFile(abs)
        return { ok: false, output: r.ok ? '' : r.msg }
      }
      const lines = raw.split('\n')
      const start = Math.max(1, Number(args.start_line) || 1)
      const end = Math.min(lines.length, Number(args.end_line) || lines.length)
      const slice = lines.slice(start - 1, end)
      const numbered = slice.map((l, i) => `${start + i}|${l}`).join('\n')
      return { ok: true, output: numbered, filePath: abs }
    }

    case 'edit_file': {
      const abs = resolvePath(ctx, String(args.path ?? ''))
      if (!abs) return { ok: false, output: 'Invalid path' }
      const r = readTextFile(abs)
      if (!r.ok) return { ok: false, output: r.msg }
      const oldStr = String(args.old_string ?? '')
      const newStr = String(args.new_string ?? '')
      const count = r.content.split(oldStr).length - 1
      if (count === 0) return { ok: false, output: 'old_string not found in file' }
      if (count > 1) return { ok: false, output: `old_string matches ${count} times — must be unique` }
      const updated = r.content.replace(oldStr, newStr)
      fs.writeFileSync(abs, updated, 'utf8')
      ctx.openFileContents.set(abs, updated)
      return {
        ok: true,
        output: `Edited ${args.path}`,
        diff: computeLineDiff(r.content, updated),
        filePath: abs,
        filesChanged: true,
      }
    }

    case 'write_file': {
      const abs = resolvePath(ctx, String(args.path ?? ''))
      if (!abs) return { ok: false, output: 'Invalid path' }
      const content = String(args.content ?? '')
      const before = fs.existsSync(abs) ? readTextFile(abs) : null
      const oldContent = before?.ok ? before.content : ''
      fs.mkdirSync(path.dirname(abs), { recursive: true })
      fs.writeFileSync(abs, content, 'utf8')
      ctx.openFileContents.set(abs, content)
      return {
        ok: true,
        output: `Wrote ${args.path}`,
        diff: oldContent !== content ? computeLineDiff(oldContent, content) : undefined,
        filePath: abs,
        filesChanged: true,
      }
    }

    case 'list_directory': {
      const rel = String(args.path ?? '.')
      const abs = resolvePath(ctx, rel === '.' ? '.' : rel)
      if (!abs || !fs.existsSync(abs)) return { ok: false, output: 'Directory not found' }
      const entries = fs.readdirSync(abs, { withFileTypes: true })
        .filter(e => !e.name.startsWith('.'))
        .map(e => `${e.isDirectory() ? '📁' : '📄'} ${e.name}`)
      return { ok: true, output: entries.join('\n') || '(empty)' }
    }

    case 'search_codebase': {
      if (!ctx.projectPath) return { ok: false, output: 'No project open' }
      const pattern = String(args.pattern ?? '')
      const ext = args.glob ? String(args.glob).replace(/^\./, '') : null
      let re: RegExp
      try { re = new RegExp(pattern, 'gi') } catch {
        return { ok: false, output: 'Invalid regex pattern' }
      }
      const hits: string[] = []
      for (const file of walkDir(ctx.projectPath)) {
        if (ext && !file.endsWith(`.${ext}`)) continue
        const r = readTextFile(file)
        if (!r.ok) continue
        const rel = path.relative(ctx.projectPath, file)
        r.content.split('\n').forEach((line, i) => {
          if (re.test(line)) {
            hits.push(`${rel}:${i + 1}: ${line.trim().slice(0, 120)}`)
            re.lastIndex = 0
          }
        })
        if (hits.length >= 40) break
      }
      return { ok: true, output: hits.join('\n') || 'No matches' }
    }

    case 'summarize_file': {
      const abs = resolvePath(ctx, String(args.path ?? ''))
      if (!abs) return { ok: false, output: 'Invalid path' }
      const r = readTextFile(abs)
      if (!r.ok) return { ok: false, output: r.msg }
      const lines = r.content.split('\n')
      const preview = lines.slice(0, 30).map((l, i) => `${i + 1}|${l}`).join('\n')
      const summary = [
        `Path: ${args.path}`,
        `Lines: ${lines.length}`,
        `Chars: ${r.content.length}`,
        `Preview (first 30 lines):`,
        preview,
        lines.length > 30 ? `\n... ${lines.length - 30} more lines` : '',
      ].join('\n')
      return { ok: true, output: summary, filePath: abs }
    }

    case 'web_search': {
      const out = await webSearch(String(args.query ?? ''))
      return { ok: true, output: out }
    }

    default:
      return { ok: false, output: `Unknown tool: ${name}` }
  }
}

File: src-tauri/src/agent/tools.rs
use crate::agent::diff::{compute_line_diff, DiffLine};
use regex::Regex;
use std::collections::HashMap;
use std::path::Path;

pub struct ToolContext {
    pub project_path: Option<String>,
    pub open_file_contents: HashMap<String, String>,
}

pub struct ToolResult {
    pub ok: bool,
    pub output: String,
    pub diff: Option<Vec<DiffLine>>,
    pub file_path: Option<String>,
    pub files_changed: bool,
}

impl ToolResult {
    fn ok(output: impl Into<String>) -> Self {
        Self { ok: true, output: output.into(), diff: None, file_path: None, files_changed: false }
    }
    fn err(output: impl Into<String>) -> Self {
        Self { ok: false, output: output.into(), diff: None, file_path: None, files_changed: false }
    }
}

fn resolve_path(ctx: &ToolContext, file_path: &str) -> Option<String> {
    let root = ctx.project_path.as_deref()?;
    if file_path == "." || file_path == "./" {
        return Some(root.to_string());
    }
    let p = Path::new(file_path);
    if p.is_absolute() {
        let norm = p.to_string_lossy().to_string();
        if norm.starts_with(root) { Some(norm) } else { None }
    } else {
        Some(
            Path::new(root)
                .join(file_path)
                .to_string_lossy()
                .to_string(),
        )
    }
}

fn read_text_file(abs: &str) -> Result<String, String> {
    let buf = std::fs::read(abs).map_err(|e| e.to_string())?;
    if buf.contains(&0) {
        return Err("Binary file — cannot read as text".to_string());
    }
    Ok(String::from_utf8_lossy(&buf).into_owned())
}

fn walk_dir(dir: &str, out: &mut Vec<String>) {
    if let Ok(rd) = std::fs::read_dir(dir) {
        for entry in rd.flatten() {
            let name = entry.file_name().to_string_lossy().to_string();
            if name.starts_with('.') || name == "node_modules" {
                continue;
            }
            let path = entry.path().to_string_lossy().to_string();
            if entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                walk_dir(&path, out);
            } else {
                out.push(path);
            }
        }
    }
}

async fn web_search(query: &str) -> String {
    let url = format!(
        "https://api.duckduckgo.com/?q={}&format=json&no_redirect=1",
        urlencoding::encode(query)
    );
    let client = reqwest::Client::new();
    let Ok(res) = client
        .get(&url)
        .header("User-Agent", "CelestiaAgent/1.0")
        .send()
        .await
    else {
        return "Search failed".to_string();
    };
    let Ok(data) = res.json::<serde_json::Value>().await else {
        return "Search parse failed".to_string();
    };

    let mut parts: Vec<String> = Vec::new();
    if let Some(t) = data["AbstractText"].as_str().filter(|s| !s.is_empty()) {
        parts.push(t.to_string());
        if let Some(u) = data["AbstractURL"].as_str() {
            parts.push(u.to_string());
        }
    }
    if let Some(topics) = data["RelatedTopics"].as_array() {
        for t in topics.iter().take(5) {
            if let Some(text) = t["Text"].as_str() {
                parts.push(format!("• {text}"));
            }
        }
    }
    if parts.is_empty() { "No results found.".to_string() } else { parts.join("\n") }
}

fn str_val(args: &HashMap<String, serde_json::Value>, key: &str) -> String {
    args.get(key)
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string()
}

pub async fn execute_tool(
    name: &str,
    args: &HashMap<String, serde_json::Value>,
    ctx: &mut ToolContext,
) -> ToolResult {
    match name {
        "read_file" => {
            let path_arg = str_val(args, "path");
            // strip archive notation suffix after ':'
            let path_clean = path_arg.split(':').next().unwrap_or(&path_arg).to_string();
            let Some(abs) = resolve_path(ctx, &path_clean) else {
                return ToolResult::err("Invalid path or no project open");
            };

            let raw = ctx
                .open_file_contents
                .get(&abs)
                .cloned()
                .or_else(|| read_text_file(&abs).ok());

            let Some(content) = raw else {
                return ToolResult::err("File not found");
            };

            let lines: Vec<&str> = content.lines().collect();
            let start = args.get("start_line").and_then(|v| v.as_u64()).unwrap_or(1).max(1) as usize;
            let end = args.get("end_line").and_then(|v| v.as_u64()).unwrap_or(lines.len() as u64) as usize;
            let end = end.min(lines.len());

            let numbered: String = lines[start - 1..end]
                .iter()
                .enumerate()
                .map(|(i, l)| format!("{}|{}", start + i, l))
                .collect::<Vec<_>>()
                .join("\n");

            ToolResult { ok: true, output: numbered, diff: None, file_path: Some(abs), files_changed: false }
        }

        "edit_file" => {
            let path_arg = str_val(args, "path");
            let Some(abs) = resolve_path(ctx, &path_arg) else {
                return ToolResult::err("Invalid path");
            };
            let Ok(content) = read_text_file(&abs) else {
                return ToolResult::err("File not found");
            };

            let old_str = str_val(args, "old_string");
            let new_str = str_val(args, "new_string");
            let count = content.matches(&old_str as &str).count();

            if count == 0 {
                return ToolResult::err("old_string not found in file");
            }
            if count > 1 {
                return ToolResult::err(format!("old_string matches {count} times — must be unique"));
            }

            let updated = content.replacen(&old_str as &str, &new_str, 1);
            if let Err(e) = std::fs::write(&abs, &updated) {
                return ToolResult::err(e.to_string());
            }
            ctx.open_file_contents.insert(abs.clone(), updated.clone());

            ToolResult {
                ok: true,
                output: format!("Edited {path_arg}"),
                diff: Some(compute_line_diff(&content, &updated)),
                file_path: Some(abs),
                files_changed: true,
            }
        }

        "write_file" => {
            let path_arg = str_val(args, "path");
            let Some(abs) = resolve_path(ctx, &path_arg) else {
                return ToolResult::err("Invalid path");
            };
            let content = str_val(args, "content");
            let old_content = read_text_file(&abs).unwrap_or_default();

            if let Some(parent) = Path::new(&abs).parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            if let Err(e) = std::fs::write(&abs, &content) {
                return ToolResult::err(e.to_string());
            }
            ctx.open_file_contents.insert(abs.clone(), content.clone());

            ToolResult {
                ok: true,
                output: format!("Wrote {path_arg}"),
                diff: if old_content != content {
                    Some(compute_line_diff(&old_content, &content))
                } else {
                    None
                },
                file_path: Some(abs),
                files_changed: true,
            }
        }

        "list_directory" => {
            let rel = {
                let v = str_val(args, "path");
                if v.is_empty() { ".".to_string() } else { v }
            };
            let Some(abs) = resolve_path(ctx, &rel) else {
                return ToolResult::err("Directory not found");
            };
            let Ok(rd) = std::fs::read_dir(&abs) else {
                return ToolResult::err("Directory not found");
            };
            let mut entries: Vec<String> = rd
                .flatten()
                .filter(|e| !e.file_name().to_string_lossy().starts_with('.'))
                .map(|e| {
                    let is_dir = e.file_type().map(|t| t.is_dir()).unwrap_or(false);
                    format!("{} {}", if is_dir { "📁" } else { "📄" }, e.file_name().to_string_lossy())
                })
                .collect();
            entries.sort();
            ToolResult::ok(if entries.is_empty() { "(empty)".into() } else { entries.join("\n") })
        }

        "search_codebase" => {
            let Some(root) = ctx.project_path.clone() else {
                return ToolResult::err("No project open");
            };
            let pattern = str_val(args, "pattern");
            let ext_filter = {
                let g = str_val(args, "glob");
                if g.is_empty() { None } else { Some(g.trim_start_matches('.').to_string()) }
            };

            let Ok(re) = Regex::new(&format!("(?i){pattern}")) else {
                return ToolResult::err("Invalid regex pattern");
            };

            let mut files = Vec::new();
            walk_dir(&root, &mut files);

            let mut hits = Vec::new();
            'outer: for file in files {
                if let Some(ref ext) = ext_filter {
                    if !file.ends_with(&format!(".{ext}")) {
                        continue;
                    }
                }
                let Ok(content) = read_text_file(&file) else { continue };
                let rel = file.strip_prefix(&root).unwrap_or(&file).trim_start_matches(['/', '\\']);
                for (i, line) in content.lines().enumerate() {
                    if re.is_match(line) {
                        hits.push(format!("{}:{}: {}", rel, i + 1, &line.trim()[..line.trim().len().min(120)]));
                        if hits.len() >= 40 {
                            break 'outer;
                        }
                    }
                }
            }
            ToolResult::ok(if hits.is_empty() { "No matches".into() } else { hits.join("\n") })
        }

        "summarize_file" => {
            let path_arg = str_val(args, "path");
            let Some(abs) = resolve_path(ctx, &path_arg) else {
                return ToolResult::err("Invalid path");
            };
            let Ok(content) = read_text_file(&abs) else {
                return ToolResult::err("File not found");
            };
            let lines: Vec<&str> = content.lines().collect();
            let preview: String = lines
                .iter()
                .take(30)
                .enumerate()
                .map(|(i, l)| format!("{}|{}", i + 1, l))
                .collect::<Vec<_>>()
                .join("\n");
            let summary = format!(
                "Path: {path_arg}\nLines: {}\nChars: {}\nPreview (first 30 lines):\n{preview}{}",
                lines.len(),
                content.len(),
                if lines.len() > 30 { format!("\n... {} more lines", lines.len() - 30) } else { String::new() }
            );
            ToolResult { ok: true, output: summary, diff: None, file_path: Some(abs), files_changed: false }
        }

        "web_search" => {
            let query = str_val(args, "query");
            let result = web_search(&query).await;
            ToolResult::ok(result)
        }

        _ => ToolResult::err(format!("Unknown tool: {name}")),
    }
}
